vembrane/__init__.py

```diff
@@ -49,52 +49,91 @@
 
 
 class Sample:
-    def __init__(self, record_idx: int, sample: str, format_data: Dict[str, Any]):
+    def __init__(self, record_idx: int, sample_name: str, sample):
         self._record_idx = record_idx
+        self._sample_name = sample_name
         self._sample = sample
-        self._data = format_data
+        self._data = {}
 
     def __getitem__(self, item):
         try:
             return self._data[item]
-        except KeyError as ke:
-            raise UnknownFormatField(self._record_idx, self._sample, ke)
+        except KeyError:
+            try:
+                sample_format = self._sample[item]
+            except KeyError as ke:
+                raise UnknownFormatField(self._record_idx, self._sample_name, ke)
+            self._data[item] = sample_format
+            return sample_format
 
 
 class Format:
-    def __init__(self, record_idx: int, sample_formats: Dict[Sample, Dict[str, Any]]):
+    def __init__(self, record_idx: int, record_samples):
         self._record_idx = record_idx
-        self._sample_formats = sample_formats
+        self._record_samples = record_samples
+        self._sample_formats = {}
 
     def __getitem__(self, item):
         try:
             return self._sample_formats[item]
-        except KeyError as ke:
-            raise UnknownSample(self._record_idx, ke)
+        except KeyError:
+            try:
+                record_sample = self._record_samples[item]
+            except KeyError as ke:
+                raise UnknownSample(self._record_idx, ke)
+            sample = Sample(self._record_idx, item, record_sample)
+            self._sample_formats[item] = sample
+            return sample
```

Comment:
Come to think of it, I'd like to switch order of FORMAT access to match other APIs, i.e. `FORMAT[format_field][sample]`

Reply:
Yep, we can do this. In a subsequent PR to not conflate things?